_posts/2023-02-03_medical-imaging-with-nix/medical-imaging-with-nix.Rmd

```diff
@@ -0,0 +1,233 @@
+date: 2023-01-27
+output:
+  distill::distill_article:
+    self_contained: false
+categories:
+  - infrastructure
+  - deployment-environments
+  - nix
+draft: false
+preview: future-compute-v2.jpg
+---
+
+
+I've been thinking for a long time about how to build software environments that are resilient to the
+ravages of time and that others can pick up and use effectively at their leisure. So when in the course of my work
+at DSAA I needed to develop an effective environment for doing machine learning on medical imaging data, I jumped at
+the chance. The fruit of that labour is:
+
+https://github.com/LKS-CHART/medical-imaging-nix
+
+a working environment of software, defined in software. It allows me or anyone else to reproduce my work or use the
+exact software computing environment I use. What do I mean by environment? In this case I mean isolated collections of 
+programs, packages, and libraries that can be used together, but don't interact with the rest of your software. 
+The concept might be familiar if you've worked with [renv](https://posit.co/blog/renv-project-environments-for-r/), [conda](https://docs.conda.io/en/latest/), or [lmod](https://lmod.readthedocs.io/en/latest/).
+
+The project is powered by [nix](https://nixos.org/), a futuristic technology gaining a foothold among certain groups of 
+software developers and computing enthusiasts. Renowned for its ability to almost fully deterministically and reproducibly 
+build software, enabling strongly isolated environments.
+
+## What is medical-imaging-nix?
+
+The repo above is a [nix](https://nixos.org/) flake that generates a software environment that can be used as a 
+starting point for future projects. 
+
+The environment contains R, python, jupyter, a suite of tools for working with medical imaging 
+files, pytorch and tensorflow with GPU support enabled. All^[well, most anyway] the things a medical imaging data 
+scientist needs to hit the ground  running on a new project. It can be forked and expanded to include other kinds of
+dependencies, you could add julia, rust, fortran, and many others with just a few lines of code.
+
+The environment can be compiled into a [apptainer container](https://apptainer.org/)^[I will probably switch to or add docker before long] for 
+portability to places that don't have nix. To get started with the environment you need nix installed with flakes 
+enabled but a new project can be initialized as simply as:
+
+```sh
+mkdir my-awesome-project
+cd my-awesome-project
+nix flake init -t github:LKS-CHART/medical-imaging-nix
+nix develop
+```
+
+the first time you run this it will take a very long time, because it will build most of the software universe from
+scratch for you^[including pytorch and tensorflow, let it run overnight.]. But after that all subsequent calls
+to nix develop will quickly drop you into your shiny new environment. If you're ready to package up your environment into
+a container you can run
+
+```sh
+nix build
```

Comment:
another thing is that building a container is a somewhat weird output of `nix build` with the default target - normally you'd expect this to build your software/analysis/whatever.

but, on that note, I think we need to understand https://zimbatm.com/notes/nixpkgs-unfree, although not necessarily before publishing this post

Reply:
why is a container not a satisfactory build target? I seem to remember someone working on encapsulating entire projects as nix builds, but I haven't gone that far. This project is to create a devShell or container, so the build making the container makes sense to me

Reply:
It's just a bit unusual (I would expect `nix build` with no additional arguments just to build all the software and locate it in `result/`).

Reply:
On second thought, if the project were called `medical-imaging-docker` I'd expect it to build a docker container (but if you want docker and apptainer builds on equal footing it's already weird and just building seems reasonable).